src/ProjectIdProvider.php

```diff
@@ -0,0 +1,206 @@
+     *
+     * @return string The current project's id.
+     * @throws DomainException
+     */
+    public function getProjectId()
+    {
+        // List of sources for determining a project id
+        $sources = [
+            [ $this, 'fromKeyFilePath'],
+            [ $this, 'fromApplicationDefaultCredentials' ],
+            [ $this, 'fromSdk' ],
+            [ $this, 'fromAppEngineStandard' ],
+            [ $this, 'fromComputeEngineMetaData' ],
+        ];
+
+        // Look through all of the sources until we find a project id
+        foreach ($sources as $callable) {
+            $id = $callable();
+            if ($id !== null) {
+                return $id;
+            }
+        }
+
+        // If a project id could not be found, throw an exception
+        throw new DomainException('Could not determine project id');
+    }
+
+    /**
+     * Check for a project id in the Key File specified in the GOOGLE_APPLICATION_CREDENTIALS
+     * environment variable
+     *
+     * @return string|null
+     */
+    private function fromKeyFilePath()
+    {
+        $credentials = CredentialsLoader::fromEnv();
+
+        if (!isset($credentials['project_id'])) {
+            return null;
+        }
+
+        return $credentials['project_id'];
+    }
+
+    /**
+     * Check for a project id in the application default credentials
+     *
+     * @return string|null
+     */
+    private function fromApplicationDefaultCredentials()
+    {
+        $credentials = CredentialsLoader::fromWellKnownFile();
+
+        if (!isset($credentials['project_id'])) {
+            return null;
+        }
+
+        return $credentials['project_id'];
+    }
+
+    /**
+     * Attempts to retrieve a project id from the gcloud sdk
+     *
+     * @return string|null
+     */
+    private function fromSdk()
+    {
+        $command = 'gcloud config config-helper --format json';
+
+        // exec the command
+        $output = $this->execute($command);
+
+        // if we didn't receive output, return null
+        if (!$output) {
+            return null;
+        }
+
+        // the output should be json. decode it and check for errors
+        $config = json_decode($output, true);
+
+        if (json_last_error() != JSON_ERROR_NONE) {
+            return null;
+        }
+
+        // the configuration should contain a project id at the specified path
+        if (!isset($config['configuration']['properties']['core']['project'])) {
+            return null;
+        }
+
+        return $config['configuration']['properties']['core']['project'];
+    }
+
+    /**
+     * Check for the project id available in App Engine Standard Environments
+     *
+     * @return string|null
+     */
+    private function fromAppEngineStandard()
+    {
+        // app engine should contain the app id environment variable
+        $appId = getenv('APPLICATION_ID');
+
+        // if we didn't find it, return null
+        if (!$appId) {
```

Comment:
How about using `AppIdentityCredentials::onAppEngine()` instead?

Reply:
I didn't apply this change in the last revision. I needed the env var anyway to parse the app id. Do you think I should add onAppEngine as an initial check?

Reply:
`APPLICATION_ID` seems rather generic; would this behave properly if someone were to set it themselves in a form different from the way App Engine uses it?

Reply:
You're right, it would definitely be wonky. I added the App Engine check!